File: chaincode/src/model/energyAccount.ts
/*
 * SPDX-License-Identifier: Apache-2.0
 */
import * as Yup from 'yup';

/* tslint:disable:max-classes-per-file */

export class EnergyAccountPoint {
    public inQuantity: number;
    public position: number;
}

export class EnergyAccount {

    public static readonly schema = Yup.object().shape({
        areaDomain: Yup.string().required('areaDomain is a compulsory string').typeError('areaDomain is a compulsory string'),
        businessType: Yup.string().notRequired(),
        classificationType: Yup.string().notRequired(),
        createdDateTime: Yup.string().required(),
        docStatus: Yup.string().notRequired(),
        docType: Yup.string().notRequired(),
        energyAccountMarketDocumentMrid: Yup.string().required(
            'energyAccountMarketDocumentMrid is a compulsory string',
        ),
        marketEvaluationPointMrid: Yup.string().notRequired(),
        measurementUnitName: Yup.string().required('measurementUnitName is a compulsory string'),
        meteringPointMrid: Yup.string().required(
            'meteringPointMrid is a compulsory string',
        ),
        processType: Yup.string().notRequired(),
        product: Yup.string().notRequired(),
        receiverMarketParticipantMrid: Yup.string().required(
            'receiverMarketParticipantMrid is a compulsory string').typeError('receiverMarketParticipantMrid must be a string'),
        receiverMarketParticipantRole: Yup.string().required(
            'receiverMarketParticipantRole is a compulsory string').typeError('receiverMarketParticipantRole must be a string'),
        resolution: Yup.string().required('resolution is a compulsory string'),
        revisionNumber: Yup.string().notRequired(),
        senderMarketParticipantMrid: Yup.string().required(
            'senderMarketParticipantMrid is a compulsory string').typeError('senderMarketParticipantMrid must be a string'),
        senderMarketParticipantRole: Yup.string().required(
            'senderMarketParticipantRole is a compulsory string',
        ),
        timeInterval: Yup.string().required(),
        timeSeries: Yup.array().of(Yup.object().shape(
            {
                inQuantity: Yup.number().required(),
                position: Yup.number().required(),
            },
        )).required(),
    });

    public docType?: string;
    public energyAccountMarketDocumentMrid: string; // PK
    public meteringPointMrid: string; // FK1
    public areaDomain: string;
    public senderMarketParticipantMrid: string;
    public senderMarketParticipantRole: string;
    public receiverMarketParticipantMrid: string;
    public receiverMarketParticipantRole: string;
    public createdDateTime: string;
    public measurementUnitName: string;
    public timeInterval: string;
    public resolution: string;
    public timeSeries: EnergyAccountPoint[];
    public startCreatedDateTime: string;
    public endCreatedDateTime: string;
    public revisionNumber?: string;
    public businessType?: string;
    public docStatus?: string;
    public marketEvaluationPointMrid?: string; // FK2
    public processType?: string;
    public classificationType?: string;
    public product?: string;
}

File: backend/service/src/main/java/com/star/models/energyaccount/EnergyAccount.java
package com.star.models.energyaccount;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.star.models.common.ValidationRegex;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import java.util.List;

/**
 * Copyright (c) 2022, Enedis (https://www.enedis.fr), RTE (http://www.rte-france.com)
 * SPDX-License-Identifier: Apache-2.0
 */
@Builder
@Data
@NoArgsConstructor
@AllArgsConstructor
public class EnergyAccount {

    @JsonIgnore
    private String docType;
    @NotBlank(message = "Le champ energyAccountMarketDocumentMrid est obligatoire")
    private String energyAccountMarketDocumentMrid;
    @NotBlank(message = "Le champ meteringPointMrid est obligatoire")
    private String meteringPointMrid;
    private String marketEvaluationPointMrid;
    private String revisionNumber;
    private String businessType;
    private String docStatus;
    private String processType;
    private String classificationType;
    private String product;
    @NotBlank(message = "Le champ areaDomain est obligatoire")
    private String areaDomain;
    @NotBlank(message = "Le champ receiverMarketParticipantMrid est obligatoire")
    private String receiverMarketParticipantMrid;
    @NotBlank(message = "Le champ receiverMarketParticipantRole est obligatoire")
    private String receiverMarketParticipantRole;
    @NotBlank(message = "Le champ senderMarketParticipantMrid est obligatoire")
    private String senderMarketParticipantMrid;
    @NotBlank(message = "Le champ senderMarketParticipantRole est obligatoire")
    private String senderMarketParticipantRole;
    @NotBlank(message = "Le champ createdDateTime est obligatoire")
    @Pattern(regexp = ValidationRegex.DATETIME_REGEX, message = "Le champ createdDateTime doit être au format : " + ValidationRegex.DATETIME_REGEX_STR)
    private String createdDateTime;
    @NotBlank(message = "Le champ measurementUnitName est obligatoire")
    private String measurementUnitName;
    @Pattern(regexp = ValidationRegex.RESOLUTION_REGEX, message = "Le champ resolution doit être au format : " + ValidationRegex.RESOLUTION_REGEX_STR)
    private String resolution;
    @NotBlank(message = "Le champ timeInterval est obligatoire")
    @Pattern(regexp = ValidationRegex.DATETIME_INTERVAL_REGEX, message = "Le champ timeInterval doit être au format : " + ValidationRegex.DATETIME_INTERVAL_REGEX_STR)
    private String timeInterval;
    private String startCreatedDateTime;
    private String endCreatedDateTime;
    @NotNull(message = "Le champ timeSeries est obligatoire")
    @NotEmpty(message = "Le champ timeSeries ne doit pas être vide")
    private List<EnergyAccountPoint> timeSeries;

}